<a href="https://colab.research.google.com/github/number000000/CSE151A-MachineLearning-TeamProject/blob/main/Model3%5BFinal%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install ucimlrepo

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
from ucimlrepo import fetch_ucirepo
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
import tensorflow as tf

In [ ]:
rice_cammeo_and_osmancik = fetch_ucirepo(id=545)

In [ ]:
X = rice_cammeo_and_osmancik.data.features
y = rice_cammeo_and_osmancik.data.targets

In [ ]:
X.info()
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3810 entries, 0 to 3809
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Area               3810 non-null   int64  
 1   Perimeter          3810 non-null   float64
 2   Major_Axis_Length  3810 non-null   float64
 3   Minor_Axis_Length  3810 non-null   float64
 4   Eccentricity       3810 non-null   float64
 5   Convex_Area        3810 non-null   int64  
 6   Extent             3810 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 208.5 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3810 entries, 0 to 3809
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Class   3810 non-null   object
dtypes: object(1)
memory usage: 29.9+ KB


In [ ]:
X.describe()

Area    Perimeter  Major_Axis_Length  Minor_Axis_Length  \
count   3810.000000  3810.000000        3810.000000        3810.000000   
mean   12667.727559   454.239180         188.776222          86.313750   
std     1732.367706    35.597081          17.448679           5.729817   
min     7551.000000   359.100006         145.264465          59.532406   
25%    11370.500000   426.144752         174.353855          82.731695   
50%    12421.500000   448.852493         185.810059          86.434647   
75%    13950.000000   483.683746         203.550438          90.143677   
max    18913.000000   548.445984         239.010498         107.542450   

       Eccentricity   Convex_Area       Extent  
count   3810.000000   3810.000000  3810.000000  
mean       0.886871  12952.496850     0.661934  
std        0.020818   1776.972042     0.077239  
min        0.777233   7723.000000     0.497413  
25%        0.872402  11626.250000     0.598862  
50%        0.889050  12706.500000     0.645361  
75%        0.902588  14284.000000     0.726562  
max        0.948007  19099.000000     0.861050

In [ ]:
X.isnull().sum()

Area                 0
Perimeter            0
Major_Axis_Length    0
Minor_Axis_Length    0
Eccentricity         0
Convex_Area          0
Extent               0
dtype: int64

In [ ]:
y['Class'].value_counts()

Osmancik    2180
Cammeo      1630
Name: Class, dtype: int64

In [ ]:
scaler = MinMaxScaler()
X_normalized = scaler.fit_transform(X)
X_normalized

array([[0.67593733, 0.87923163, 0.90121587, ..., 0.88801055, 0.69391702,
        0.20757716],
       [0.62533005, 0.71409491, 0.64808716, ..., 0.69197987, 0.64600914,
        0.32456423],
       [0.62339377, 0.75006612, 0.73434911, ..., 0.78984635, 0.63563643,
        0.53857594],
       ...,
       [0.43002992, 0.44045819, 0.40864083, ..., 0.60847407, 0.43266526,
        0.35564534],
       [0.20515754, 0.17532451, 0.16992207, ..., 0.56901705, 0.20868495,
        0.44453926],
       [0.34175321, 0.24088172, 0.16869838, ..., 0.28376388, 0.34001406,
        0.84022485]])

In [ ]:
label_encoder = LabelEncoder()
y['Class'] = label_encoder.fit_transform(y['Class'])

<ipython-input-69-d20a001ba4a5>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y['Class'] = label_encoder.fit_transform(y['Class'])


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.2, random_state=0)

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import CategoricalNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, log_loss


knn = KNeighborsClassifier()

knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)


accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
loss = log_loss(y_test, y_pred)
test_error = 1-accuracy

print(f'Accuracy: {accuracy}')
print('Classification Report:\n', report)
print('Loss:\n', loss)
print('Test Error:\n', test_error)

Accuracy: 0.926509186351706
Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.90      0.92       337
           1       0.92      0.95      0.94       425

    accuracy                           0.93       762
   macro avg       0.93      0.92      0.93       762
weighted avg       0.93      0.93      0.93       762

Loss:
 2.648877414423308
Test Error:
 0.07349081364829402


/usr/local/lib/python3.10/dist-packages/sklearn/neighbors/_classification.py:215: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


# We mordify the code to use KNN

Originally, we want to use Naive Bayes for model 3, but we found out that the metrics of this model are very bad, so we decide to change the model to KNN, which perform very well

## Evaluate your data, labels and loss function. Were they sufficient or did you have have to change them.

The data and label is the same as Milestone 4. The result in Milestone 4 is good enough so we decide not to change that.

The loss we use to evelate the model is log loss.

## Evaluate your model compare training vs test error
testing error: 0.07349081364829402

$$ \frac{FP+FN}{TP+FP+TN+FN} = 1 - \text{accuracy} $$

## Where does your model fit in the fitting graph.
Accuracy, Precision, Recall, and F1-score are consistently high on the test dataset (around 93%), indicating that the model performs well on unseen data.


We can see the training and testing metrics perform good and similar to each other which show that our model is just right. No overfitting or underfitting happen.

It is about the same as the first model. Since our first model is already have good performance, this model is consider good.

## Conclusion section

The model currently perform well, with above 93% accuracy on testing data. The model's test accuracy is a bit better than the previous one, which show that KNN can perform better in this classification task.